tests/jvm/src/test/scala/munit/AsyncFixtureOrderSuite.scala

```diff
@@ -0,0 +1,38 @@
+package munit
+
+import scala.concurrent.Future
+import scala.concurrent.Promise
+
+class AsyncFixtureOrderSuite extends FunSuite {
+  val latch = Promise[Unit]
+  var completedFromTest = Option.empty[Boolean]
+  var completedFromTeardown = Option.empty[Boolean]
+
+  val latchOnTeardown = FunFixture.async[String](
+    setup = { test => Future.successful(test.name) },
```

Comment:
If you are interested, we can take this discussion offline, detailed concurrency discussions can annoy others. Or, if you think I am generating more noise than benefit,
I can butt out.

I think this line is the crux.  I would have to trace what in the environment is supposed to be sequencing setup, test, & teardown.  Is the test on this line the one which eventually gets executed or does something else?   I strongly suspect that the test => body
should be using either inline code in the .successful() or a .onComplete (or .andThen)
to complete the Promise (data type of the Promise value could change to Boolean).

Reply:
I spent a long day tracing code and also discovered and read munit/doc. Agreed, the latter was unfair.

I think I am right in the concept that the test should be setting the latch (or failing if already set) and the teardown checking that the latch is set (and failing or throwing an exception if not). I was wrong about suggesting that the latch be set in setup. Sorry.

My understanding is increasing, but as of now it looks to me if the Thread.sleep() is needed here the code is detecting a valid problem.  Visual inspection of the chain of
completion of Futures looked pretty good to me. Of course, that is why executable tests
exist.

Reply:
As the code currently stands, `Thread.sleep()` is not needed. That is because the order of the execution of test and teardown is controlled - teardown executes only after test completed. This is guaranteed by the following lines:

https://github.com/scalameta/munit/pull/91/files#diff-64b49405259d255602bf67e08b1c301fR33-R38

But if, in the future this code gets refactored, to lets say:

```scala
val b = munitValueTransform(body(argument))
teardown(argument)
b
```

The test without the `Thread.sleep()` would still mostly pass. By adding `Thread.sleep()` to the test body, the test makes sure that enough time is spent in the test body and teardown is given a chance to execute, if it wants.

Reply:
1) Did you get Thread.sleep() to link on SN on MacOS?   

2) Yes, that is the code I discovered. I think we both agree that the refactoring of the code such as you described is a blunder of the first magnitude.  IMHO, even using the Haskell style 
approach to comments that section code deserves a "Curse of the Mummy" comment, or at least a skull & crossbones, describing the assumptions and or pre and post-conditions.
Copying and slightly enhancing the description of "happens-before" to that section of 
code would make it clear at the point of tinkering what needed to be preserved
(yeah, it should be obvious to anyone reading the code but it might give 
a pause for reflection to senior developers in a hurry or junior devos. Certainly
it would be a clue for reviewers).  I'm trying to get to root causes to your well
founded concern. The code at the point of weakness is within your control, not arbitrary
 user code. Making it harder, abeit not impossible, for people to blunder and void the
pre- & post-conditions is, IMHO, the primary leverage point, with testing being the 
secondary.

One can not test in quality, but we both agree testing helps.
As your effort shows, comments can not prevent foolishness. Runtime tests sanity checks
are good, if they are believable.  Even with Thread.sleep(), the test, as is, remains stochastic.
Agreed, the longer the sleep, the longer the recurrence period of the Heisenbug.  
Does _every_ run of the munit test ensemble need to be delayed even a fraction of a 
second for questionable benefit, and that only on JVM? 

Some test frameworks have a concept of repeating a test a fixed number of times, That concept helps catch Heisenbugs. 

I propose that this PR proceeds. After it is committed and if I have the time, I can submit
a PR to AsyncFixtureOrderSuite.scala on top of this one.  That would give a concrete instance of what I am proposing.  Easier to review and it would not offend me if it were never accepted.

Reply:
The mention of "happens-before" in the code comment got me thinking about
pre- and post-conditions.  In particular AsyncFixtureOrderSuite checks the condition
that teardown executes after the test. I believe did not see code which tests if setup executes
before test.  I beg pardon if it is there and I did not see it. 

 Does that condition get tested some place else in the ensemble?  I would expect to 
see it in the referenced file, or else a one line comment as to why it was missing.

My comments are meant to help the effort, not be picky or wast munit devo time.

Reply:
> Did you get Thread.sleep() to link on SN on MacOS?

Nope, that was my original problem. :) I currently worked it around by moving the test that uses `Thread.sleep()` to a JVM-only test.

> Copying and slightly enhancing the description of "happens-before" to that section of
code would make it clear at the point of tinkering

Sounds good. I'll add a short comment. But I always prefer tests to writing comments, because one can not run comments. :)